File: tests/tooling/perf/nbody.c
// The five-body simulation from the benchmark game: pairwise force
// accumulation over a small struct array, dominated by sqrt and by the
// dependency chain through each body's velocity. Float-heavy in a way the
// array benchmarks are not — they stream, this one recurses on its own
// results.
#include <stdio.h>
#include <math.h>

#define NB 5
#define SOLAR_MASS (4 * 3.141592653589793 * 3.141592653589793)
#define DAYS_PER_YEAR 365.24

typedef struct { double x, y, z, vx, vy, vz, mass; } body;

static body bodies[NB] = {
    { 0, 0, 0, 0, 0, 0, SOLAR_MASS },
    { 4.84143144246472090e+00, -1.16032004402742839e+00, -1.03622044471123109e-01,
      1.66007664274403694e-03 * DAYS_PER_YEAR, 7.69901118419740425e-03 * DAYS_PER_YEAR,
      -6.90460016972063023e-05 * DAYS_PER_YEAR, 9.54791938424326609e-04 * SOLAR_MASS },
    { 8.34336671824457987e+00, 4.12479856412430479e+00, -4.03523417114321381e-01,
      -2.76742510726862411e-03 * DAYS_PER_YEAR, 4.99852801234917238e-03 * DAYS_PER_YEAR,
      2.30417297573763929e-05 * DAYS_PER_YEAR, 2.85885980666130812e-04 * SOLAR_MASS },
    { 1.28943695621391310e+01, -1.51111514016986312e+01, -2.23307578892655734e-01,
      2.96460137564761618e-03 * DAYS_PER_YEAR, 2.37847173959480950e-03 * DAYS_PER_YEAR,
      -2.96589568540237556e-05 * DAYS_PER_YEAR, 4.36624404335156298e-05 * SOLAR_MASS },
    { 1.53796971148509165e+01, -2.59193146099879641e+01, 1.79258772950371181e-01,
      2.68067772490389322e-03 * DAYS_PER_YEAR, 1.62824170038242295e-03 * DAYS_PER_YEAR,
      -9.51592254519715870e-05 * DAYS_PER_YEAR, 5.15138902046611451e-05 * SOLAR_MASS }
};

static void advance(double dt) {
    for (int i = 0; i < NB; i++)
        for (int j = i + 1; j < NB; j++) {
            double dx = bodies[i].x - bodies[j].x;
            double dy = bodies[i].y - bodies[j].y;
            double dz = bodies[i].z - bodies[j].z;
            double d2 = dx * dx + dy * dy + dz * dz;
            double mag = dt / (d2 * sqrt(d2));
            bodies[i].vx -= dx * bodies[j].mass * mag;
            bodies[i].vy -= dy * bodies[j].mass * mag;
            bodies[i].vz -= dz * bodies[j].mass * mag;
            bodies[j].vx += dx * bodies[i].mass * mag;
            bodies[j].vy += dy * bodies[i].mass * mag;
            bodies[j].vz += dz * bodies[i].mass * mag;
        }
    for (int i = 0; i < NB; i++) {
        bodies[i].x += dt * bodies[i].vx;
        bodies[i].y += dt * bodies[i].vy;
        bodies[i].z += dt * bodies[i].vz;
    }
}

static double energy(void) {
    double e = 0.0;
    for (int i = 0; i < NB; i++) {
        e += 0.5 * bodies[i].mass *
             (bodies[i].vx * bodies[i].vx + bodies[i].vy * bodies[i].vy + bodies[i].vz * bodies[i].vz);
        for (int j = i + 1; j < NB; j++) {
            double dx = bodies[i].x - bodies[j].x;
            double dy = bodies[i].y - bodies[j].y;
            double dz = bodies[i].z - bodies[j].z;
            e -= (bodies[i].mass * bodies[j].mass) / sqrt(dx * dx + dy * dy + dz * dz);
        }
    }
    return e;
}

int main(void) {
    double px = 0.0, py = 0.0, pz = 0.0;
    for (int i = 0; i < NB; i++) {
        px += bodies[i].vx * bodies[i].mass;
        py += bodies[i].vy * bodies[i].mass;
        pz += bodies[i].vz * bodies[i].mass;
    }
    bodies[0].vx = -px / SOLAR_MASS;
    bodies[0].vy = -py / SOLAR_MASS;
    bodies[0].vz = -pz / SOLAR_MASS;
    for (int i = 0; i < 3000000; i++) advance(0.01);
    printf("%.9f\n", energy());
    return 0;
}

File: tests/tooling/perf/nbody.rs
// The Rust twin of nbody.fpp: the same five-body simulation over parallel
// arrays, the same constants, the same grouping of the addends.
const PI: f64 = 3.141592653589793;
const SOLAR_MASS: f64 = 4.0 * PI * PI;
const DAYS_PER_YEAR: f64 = 365.24;
const NB: usize = 5;

fn main() {
    let mut bx = [0.0f64; NB];
    let mut by = [0.0f64; NB];
    let mut bz = [0.0f64; NB];
    let mut vx = [0.0f64; NB];
    let mut vy = [0.0f64; NB];
    let mut vz = [0.0f64; NB];
    let mut mass = [0.0f64; NB];

    mass[0] = SOLAR_MASS;

    bx[1] = 4.84143144246472090; by[1] = -1.16032004402742839; bz[1] = -0.103622044471123109;
    vx[1] = 0.00166007664274403694 * DAYS_PER_YEAR;
    vy[1] = 0.00769901118419740425 * DAYS_PER_YEAR;
    vz[1] = -0.0000690460016972063023 * DAYS_PER_YEAR;
    mass[1] = 0.000954791938424326609 * SOLAR_MASS;

    bx[2] = 8.34336671824457987; by[2] = 4.12479856412430479; bz[2] = -0.403523417114321381;
    vx[2] = -0.00276742510726862411 * DAYS_PER_YEAR;
    vy[2] = 0.00499852801234917238 * DAYS_PER_YEAR;
    vz[2] = 0.0000230417297573763929 * DAYS_PER_YEAR;
    mass[2] = 0.000285885980666130812 * SOLAR_MASS;

    bx[3] = 12.8943695621391310; by[3] = -15.1111514016986312; bz[3] = -0.223307578892655734;
    vx[3] = 0.00296460137564761618 * DAYS_PER_YEAR;
    vy[3] = 0.00237847173959480950 * DAYS_PER_YEAR;
    vz[3] = -0.0000296589568540237556 * DAYS_PER_YEAR;
    mass[3] = 0.0000436624404335156298 * SOLAR_MASS;

    bx[4] = 15.3796971148509165; by[4] = -25.9193146099879641; bz[4] = 0.179258772950371181;
    vx[4] = 0.00268067772490389322 * DAYS_PER_YEAR;
    vy[4] = 0.00162824170038242295 * DAYS_PER_YEAR;
    vz[4] = -0.0000951592254519715870 * DAYS_PER_YEAR;
    mass[4] = 0.0000515138902046611451 * SOLAR_MASS;

    let mut px = 0.0; let mut py = 0.0; let mut pz = 0.0;
    for i in 0..NB {
        px += vx[i] * mass[i];
        py += vy[i] * mass[i];
        pz += vz[i] * mass[i];
    }
    vx[0] = -px / SOLAR_MASS;
    vy[0] = -py / SOLAR_MASS;
    vz[0] = -pz / SOLAR_MASS;

    let dt = 0.01;
    for _ in 0..3000000 {
        for i in 0..NB {
            for j in (i + 1)..NB {
                let dx = bx[i] - bx[j];
                let dy = by[i] - by[j];
                let dz = bz[i] - bz[j];
                let d2 = dx * dx + dy * dy + dz * dz;
                let mag = dt / (d2 * d2.sqrt());
                vx[i] -= dx * mass[j] * mag;
                vy[i] -= dy * mass[j] * mag;
                vz[i] -= dz * mass[j] * mag;
                vx[j] += dx * mass[i] * mag;
                vy[j] += dy * mass[i] * mag;
                vz[j] += dz * mass[i] * mag;
            }
        }
        for k in 0..NB {
            bx[k] += dt * vx[k];
            by[k] += dt * vy[k];
            bz[k] += dt * vz[k];
        }
    }

    let mut e = 0.0f64;
    for i in 0..NB {
        e += 0.5 * mass[i] * (vx[i] * vx[i] + vy[i] * vy[i] + vz[i] * vz[i]);
        for j in (i + 1)..NB {
            let dx = bx[i] - bx[j];
            let dy = by[i] - by[j];
            let dz = bz[i] - bz[j];
            e -= (mass[i] * mass[j]) / (dx * dx + dy * dy + dz * dz).sqrt();
        }
    }
    println!("{:.9}", e);
}
